python/mxnet/metric.py

```diff
@@ -475,8 +475,84 @@ def update(self, labels, preds):
             self.num_inst += num_samples
 
 
+class _BinaryClassificationMixin(object):
+    """
+    Private mixin for keeping track of TPR, FPR, TNR, FNR counts for a classification metric.
+    """
+
+    def __init__(self):
+        self._true_positives = 0
+        self._false_negatives = 0
+        self._false_positives = 0
+        self._true_negatives = 0
+
+    def _update_binary_stats(self, label, pred):
+        """
+        Update various binary classification counts for a single (label, pred)
+        pair.
+
+        Parameters
+        ----------
+        label : `NDArray`
+            The labels of the data.
+
+        pred : `NDArray`
+            Predicted values.
+        """
+        pred = pred.asnumpy()
+        label = label.asnumpy().astype('int32')
+        pred_label = numpy.argmax(pred, axis=1)
+
+        check_label_shapes(label, pred)
+        if len(numpy.unique(label)) > 2:
+            raise ValueError("%s currently only supports binary classification."
+                             % self.__class__.__name__)
+
+        for y_pred, y_true in zip(pred_label, label):
```

Comment:
Do we have to use a for-loop here? Using array arithmetic OPs should be more efficient. Also we can try not to convert them into numpy array so it can be calculated in GPU.

Reply:
I agree, but there's another issue for that, so I assumed it would be done in a separate PR. https://github.com/apache/incubator-mxnet/issues/9586

Reply:
Let's address that in a separate PR. My last attempt faced performance issues when switching to ndarray-based logic, so we should address the problem once it's more clear how to resolve the performance issue.

Reply:
I've used codes like 
```python
tp = nd.sum((pred == 1) * (label == 1)).asscalar()
fp = nd.sum((pred == 1) * (label == 0)).asscalar()
fn = nd.sum((pred == 0) * (label == 1)).asscalar()
precision = float(tp) / (tp + fp)
recall = float(tp) / (tp + fn)
f1 = 2 * (precision * recall) / (precision + recall)
```
to calculate the F1 and I find it's much faster in GPU.

Reply:
Shouldn't this conversation take place on #9586? This PR only aims to fix #9588.

Reply:
OK, it should be addressed in a later PR.